### PR TITLE
Further improve CJK and encoded URL support

### PR DESCRIPTION
**Pull request by scarf005**
- added [`punycode.js`](https://www.npmjs.com/package/punycode.js) to parse punycode encoded path elements from `URL` constructor
- handles CJK path for blobs, wiki pages, and external web pages

+ ~stacked on top of #60~
+ continuation of #57
+ related: https://github.com/refined-github/refined-github/issues/7744#issuecomment-2309885325

### CODE REVIEW

File: index.test.js [hunk 1]
```diff
@@ -539,7 +548,11 @@ test('External URLs', urlMatcherMacro, new Map([
 		'https://example.site/한글로-된-URL',
 		'example.site/한글로-된-URL',
 	],
-]));
+	[
+		'https://한글로-된-경로/하위경로#한글-해시',
+		'한글로-된-경로/하위경로#한글-해시',
```
Comment (fregante) [inline index.test.js:553]: Where's the dot?

**Review by fregante — COMMENTED** [review]
Thanks! I was thinking about this while working on https://github.com/refined-github/shorten-repo-url/commit/b5d9e282903b9eeba0b002fa09feca59d6a2537d

Can you check if it fixed it? More test cases are definitely welcome so that regressions are caught. Especially for searches like:

https://github.com/refined-github/shorten-repo-url/issues?q=sort%3Aupdated-desc+is%3Aissue+is%3Aopen+label%3A%22help+wanted%22

File: index.test.js (hunk 1 above)
Comment (scarf005) [reply index.test.js]: i forgot, will add

File: __snapshots__/index.test.js.snap [hunk 2]
```diff
@@ -108,6 +110,8 @@ exports[`https://github.com/fregante/shorten-repo-url/tree/v0.12/doc 1`] = `<cod
 
 exports[`https://github.com/fregante/shorten-repo-url/wiki/%22Can-you-add-this-feature%3F%22 1`] = `Wiki: "Can you add this feature?"`;
 
+exports[`https://github.com/fregante/shorten-repo-url/wiki/한글-위키-페이지-경로 1`] = `Wiki: 한글 위키 페이지 경로`;
```
Comment (fregante) [inline __snapshots__/index.test.js.snap:113]: If you have it, it would be good to see real URLs, especially because GitHub might vary how it handles these over time.

Also can you add a test for `#hash-on-the-wiki-page` like

```
https://github.com/refined-github/refined-github/wiki/Contributing#create-a-new-feature
```

**Comment by fregante** [comment]
By the way, you don't have to force-push, it makes it difficult to see what changed between reviews

File: index.js [hunk 3]
```diff
@@ -86,11 +87,11 @@ function shortenRepoUrl(href, currentUrl = 'https://github.com') {
 	const url = new URL(href);
 	const {
 		origin,
-		pathname,
 		search,
 		searchParams,
 		hash,
 	} = url;
+	const pathname = decodeURIComponent(punycode.toUnicode(url.pathname));
```
[the remark this answers is not in the document]
Comment (scarf005) [reply index.js:94]: image

pathname seems to be punycode encoded, at least in firefox.

also, test result without pathname encoding:

```
⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯ Failed Tests 2 ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯

 FAIL  index.test.js > https://github.com/fregante/shorten-repo-url/blob/main/한글.txt
Error: Snapshot `https://github.com/fregante/shorten-repo-url/blob/main/한글.txt 1` mismatched

Expected: "<code>main</code>/한글.txt"
Received: "<code>main</code>/%ED%95%9C%EA%B8%80.txt"

 ❯ index.test.js:156:20
    154| test.each(urls)('%s', url => {
    155|  const shortened = shortenUrl(url, currentLocation);
    156|  expect(shortened).toMatchSnapshot();
       |                    ^
    157|  console.log({url});
    158|  const a = document.createElement('a');

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯[1/2]⎯

 FAIL  index.test.js > https://한글로-된-경로.com/하위경로#한글-해시
AssertionError: expected 'https://한글로-된-경로.com/하위경로#한글-해시' to be '한글로-된-경로.com/하위경로#한글-해시' // Object.is equality

Expected: "한글로-된-경로.com/하위경로#한글-해시"
Received: "https://한글로-된-경로.com/하위경로#한글-해시"

 ❯ index.test.js:167:22
    165|  applyToLink(a, currentLocation);
    166|  expect(a.href, 'applyToLink should not alter the href').toBe(correctlyEncodedUrl);
    167|  expect(a.innerHTML).toBe(shortened);
       |                      ^
    168| });
    169| 

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯[2/2]⎯

  Snapshots  1 failed
 Test Files  1 failed (1)
      Tests  2 failed | 132 passed (134)
   Start at  22:48:59
   Duration  266ms
```

**Comment by scarf005** [comment]
> By the way, you don't have to force-push, it makes it difficult to see what changed between reviews

sorry, out of habit. won't do it next time.

File: __snapshots__/index.test.js.snap (hunk 2 above)
Comment (scarf005) [reply __snapshots__/index.test.js.snap:113]: sure, would it be ok for me to use URL from my repo? (since i have no permission to edit wiki pages in refined-github)

File: index.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (fregante) [reply index.js:94]: That's not punycode, it's basic URL encoding. You can use decodeUri or decodeUriComponent

File: __snapshots__/index.test.js.snap (hunk 2 above)
Comment (fregante) [reply __snapshots__/index.test.js.snap:113]: Yes sure

File: index.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (fregante) [reply index.js:94]: I'm going to add these here for live-testing with Refined GitHub enabled:

- https://github.com/fregante/shorten-repo-url/wiki/한글-위키-페이지-경로
	- ✅ this seems to work correctly
- https://github.com/fregante/shorten-repo-url/blob/main/한글.txt
	- ❌this is decoded, I think #57 fixed it ✅
- https://한글로-된-경로.com/하위경로#한글-해시
	- ❌ this is keeping the `https://` part, but I see no issues

[the remark this answers is not in the document]
Comment (scarf005) [reply index.js]: ...oh. fixed in a4dd70c.

File: __snapshots__/index.test.js.snap (hunk 2 above)
Comment (scarf005) [reply __snapshots__/index.test.js.snap]: done in 7302dcbd5f2d92d4d6b738d0af68e162b1927408

File: index.test.js [hunk 4]
```diff
@@ -146,6 +149,7 @@ const urls = [
 	'https://www.npmjs.com/packaasdge/node',
 	'https://example.com/nodejs/node/blob/cc8fc46/.gitignore',
 	'https://example.site/한글로-된-URL',
+	'https://한글로-된-경로.com/하위경로#한글-해시',
```
Comment (fregante) [inline index.test.js:152]: I think this is failing because of this comparison, so look here:

https://github.com/refined-github/shorten-repo-url/blob/7c05f2c24eda58c4b87ea88e58b3ae1ed1be3073/index.js#L303-L312

Comment (fregante) [reply index.test.js:152]: I think that instead of using safeDecode, it should use `new URL(getLinkHref(a)).href === new URL(a.textContent).href`. This function can normalize the URL without passing it through `safeDecode`

The catch is that `new URL()` will throw when `a.textContent` is not a URL, so you still need a `safe` function.

Try just changing to this:

```js
function safeDecode(url) {
	try {
		return new URL(url).href;
	} catch {
		return url;
	}
}
```

Comment (scarf005) [reply index.test.js:152]: done in 365b3eafa4059407e18c14dac2ee559ef3dcb38f

**Comment by fregante** [comment]
This seems to be unfortunately correct, however I'm not a fan of loading that library for such a rare occurrence. However I think we have an alternative:

1. in `applyToLink`, pass the label, not the URL (`getLinkHref(a)` wasn't needed here), so we have access to the user-formatted URL rather than a pre-parsed version.
2. On line 180, I think we can drop `decodeURI` altogether
3. On line 163, we can extract the origin manually (`href.split('/')[2]`) instead of using `parsedUrl.origin`

**Review by fregante — APPROVED** [review]
Might be good now

File: index.js [hunk 5]
```diff
@@ -81,16 +81,15 @@ function shortenRepoUrl(href, currentUrl = 'https://github.com') {
...
+	// Use URL exclusively for search parameters because they're too hard to parse
 	const url = new URL(href);
-	const {
-		origin,
-		pathname,
-		search,
-		searchParams,
-		hash,
-	} = url;
+	const {search, searchParams} = url;
```
Comment (fregante) [inline index.js:92]: This part, together with the change in `applyToLink`, should resolve the problem at the root: `shortenRepoUrl()` now works directly on what the user sees, not the result of `new URL`

I'll release this as a breaking version
